sycl/include/CL/sycl/device.hpp

```diff
@@ -204,6 +208,9 @@ class __SYCL_EXPORT device {
 
   template <class T>
   friend T detail::createSyclObjFromImpl(decltype(T::impl) ImplObj);
+
+  friend device level_zero::make_device(const platform &Platform,
```

Comment:
The make_device() function must become a friend of the platform class so that it is able to access the device constructor which is currently private. Here, make_device() needs access to this constructor so that it is able to call getOrMakeDeviceImpl() (which returns a shared_pointer to a device_impl object and will be used to construct a device object). We want to call getOrMakeDeviceImpl() from within make_device() so that devices created using a Native Handle can access the cache of device_impl objects that are maintained in getOrMakeDeviceImpl(). 

Since the device constructor is private, the compiler will throw a build error complaining that the constructor "is private within this context" if make_device is not declared as a friend of the device class.

Reply:
I think we should not directly reference anything in `level_zero` namespace here.
The `platform_impl::getOrMakeDeviceImpl` implements the cache, and you just need to create the SYCL device from the return device_impl with an already friend `detail::createSyclObjFromImpl<device>`

Reply:
Understood, I have removed my changes from the device and platform headers and i have used the existing `detail::createSyclObjFromImpl` function to create the SYCL devices and platforms.